presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4

```diff
@@ -540,6 +541,28 @@ STRING
     : '\'' ( ~'\'' | '\'\'' )* '\''
     ;
 
+// note that the following is really not precisely binaryLiteral, but
+// "something that looks like binaryLiteral". here we also include
+// the form X'any string', which is a valid typeConstructor syntax
+// but we also disallow users from declaring a type 'X', so the expression
+// X'any string' is thrown out later in parsing.
+// the rationale for this is that when user writes X'some string', he's most
+// likely trying to write a binary literal, so processing the error inside
+// the binaryLiteral parsing route would give more informative error msg
+BINARY_LITERAL
+// nor should there be any WS between the starting "X" and the quote
+// but in the same spirit, we allow the WS and throw the error inside the parsing
+// for binaryLiteral
+
+// in this syntax rule, we also allow odd-number of digits, which is not exactly
+// the sql spec, but we check for that later in the parsing flow
+    :  'X' WS* '\''  WS* (BINARY_DIGIT WS*)* '\''
```

Comment:
Don't allow the first space here. Put an error message in `ExpressionAnalyzer.visitGeneralLiteral`.